### DISCUSSION TITLE
Require explicit import for React

### DISCUSSION THREAD

**Discussion by OldStarchy** [vercel/next.js#37802] [Help]
TL;DR:

I'd like `<test foo="bar" />` to compile to `myCustomCreateElement('test', {foo: 'bar'})` instead of `React.createElement('test', {foo: 'bar'})`. (but just for some files, I also need actual react to work too...)

.

I know that including

```tsx
import * as React from 'react';
```

in every file is annoying so nextjs automatically imports React ([comment](https://github.com/vercel/next.js/issues/12964#issuecomment-629642157)).

However, I'd like to try writing a custom `createElement` function for use with the JSX syntax to produce some non-html output/data structures.

I tried something like this

```tsx
namespace React {
	export function createElement(...args: any[]) {
		console.log('hello hello');
		console.log(args);
		return args;
	}
	export function render(a: any, b: any) {
		console.log(a, b);
	}
}

declare global {
	namespace JSX {
		interface IntrinsicElements {
			test: any;
		}
	}
}

export function test() {
	return React.render(<test foo="bar" />, 'test');
}
```

But as you might expect, my custom `React.createElement` function doesn't get called, and I _suspect_ it might be due to nextjs automatically using the "real" `createElement` instead of my own.

I don't know if it will work but maybe if nextjs didn't do that, then webpack might use the custom one instead (I'm not really sure) and I'd be able to use JSX syntax to do all sorts of nifty things.

Is there a way to disable the automagic react import aka require react to be explicitly imported for jsx?

Or, if anyone has any advice on how I might be able to implement this?

I considered that I might need a separate compilation step for this but I have no idea how I'd fit that into the nextjs ecosystem.

**Comment by icyJoseph** [comment]
Hi,

This was fun, long time ago I had made a ["Reactive" network with JSX, and a custom pragma](https://codesandbox.io/s/friendly-nash-wgk7x?from-embed), and this was a good excuse to go back to revisit that.

I knew that I needed to specify the runtime for JSX, and that is done with `/** @jsxRuntime classic */`, followed by the usual `/** @jsx h */` at top of the file.

Had to a bunch of type acrobatics to get it to work with TypeScript (I don't expect them to be tight enough), but here it is:

```tsx
// Reactive.tsx
/** @jsxRuntime classic */
/** @jsx h */
/** @jsxFrag frag */

type JSXElement = JSX.Element;

const Input = ({ children }: { children: number }) => <>{Number(children)}</>;

const Compute = ({
  children = [],
  compute,
}: {
  children: JSXElement | JSXElement[];
  compute: (...x: number[]) => number;
}) => (
  <>
    {Array.isArray(children)
      ? compute(...children.map(Number))
      : compute(Number(children))}
  </>
);

const Collect = ({
  children = [],
  init = 0,
}: {
  children: JSXElement[];
  init?: number;
}) => {
  return <>{children.reduce((prev, curr) => prev + Number(curr), init)}</>;
};

const Inputs = ({ children = [] }: { children: number[] }): JSXElement => (
  <>
    {Array.isArray(children) ? (
      children.map((child, index) => <Input key={index}>{child}</Input>)
    ) : (
      <Input>{children}</Input>
    )}
  </>
);

const Reactive = ({ x, y, z }: { x: number; y: number; z: number }) => (
  <Collect>
    <Compute compute={(a, b) => a + b}>
      <Inputs>{[x, z]}</Inputs>
    </Compute>
    <Compute compute={(a, b) => a * b}>
      <Compute compute={(a) => a * a * a}>
        <Input>{x}</Input>
      </Compute>
      <Compute compute={(a, b, c) => a * b * c}>
        <Inputs>{[x, y, z]}</Inputs>
      </Compute>
    </Compute>
  </Collect>
);

function frag(node: { children: Array<string | number> }) {
  return node.children;
}

type Node = (...args: unknown[]) => JSXElement;

function h(node: Node, attributes: Record<string, unknown>, ...args: Node[]) {
  let children: Node[] = args.length ? ([] as Node[]).concat(...args) : [];
  return node({ ...attributes, children });
}

export default Reactive;
```

And then we use it in a page:

```tsx
import type { NextPage } from "next";

import { useReducer } from "react";
import Reactive from "../components/Reactive";

const reducer = (state: number, action: { type: "inc" | "dec" }) => {
  switch (action.type) {
    case "inc":
      return state + 1;
    case "dec":
      return state - 1;
    default:
      return state;
  }
};

const Home: NextPage = () => {
  const [x, dispatchX] = useReducer(reducer, 0);
  const [y, dispatchY] = useReducer(reducer, 0);
  const [z, dispatchZ] = useReducer(reducer, 0);

  return (
    <div className="centered">
      <h1>Hello World</h1>
      <h2>Custom JSX driven Reactive grid!</h2>

      <div>
        <span>x</span>
        <button onClick={() => dispatchX({ type: "inc" })}>+</button>
        <button onClick={() => dispatchX({ type: "dec" })}>-</button>
      </div>

      <div>
        <span>y</span>
        <button onClick={() => dispatchY({ type: "inc" })}>+</button>
        <button onClick={() => dispatchY({ type: "dec" })}>-</button>
      </div>

      <div>
        <span>z</span>
        <button onClick={() => dispatchZ({ type: "inc" })}>+</button>
        <button onClick={() => dispatchZ({ type: "dec" })}>-</button>
      </div>

      <p>
        <Reactive x={x} y={y} z={z} />
      </p>
    </div>
  );
};

export default Home;
```

Fun stuff :D

**Reply by OldStarchy** [reply]
Wow interesting, thanks a bunch for the example code. I didn't know about those @jsx flags. I'll have to have a go at this when I get time after work.

**Reply by icyJoseph** [reply]
@OldStarchy Did you ever get to try this out?

**Reply by OldStarchy** [reply]
TL:DR; no.

I wanted to use this for two projects, though I've only worked on one, and for that, I found and have been using [React Three Fiber](https://docs.pmnd.rs/react-three-fiber/getting-started/introduction). The other project I haven't looked at in a long while.
